### ISSUE TITLE
Memory leak using nested name_scope in eager mode

### ISSUE THREAD

**Issue by matthieucoquet** [tensorflow/tensorflow#45322]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Both Windows 10 and Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: Using desktop only
- TensorFlow installed from (source or binary): tf-nightly
- TensorFlow version (use command below): 2.5.0-dev20201201
- Python version: Both 3.7 and 3.8
- Bazel version (if compiling from source): Not compiling from source
- GCC/Compiler version (if compiling from source): Not compiling from source
- CUDA/cuDNN version: 11
- GPU model and memory: RTX 2070

**Describe the current behavior**
It seems using nested name_scope can lead to memory leak. For example, in my setup, the memory grows indefinitely using this code.
```
import tensorflow as tf
for i in range(int(1e7)):
    with tf.name_scope("first"):
        with tf.name_scope("second"):
            pass
```
I'm actually not sure this is valid usage of name_scope. However, it seems this king of nested usage of name_scope is used in Tensorflow Probability ([here for example](https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/distributions/distribution.py#L1542)). Another minimal example leading to memory leak is:
```
import tensorflow_probability as tfp
distr = tfp.distributions.Normal(loc=1.0, scale=2)
for i in range(int(1e7)):
    distr.sample()
```

I don't know why, but appending "/" to the name of the scope seems to fix the leak. Here are samples without memory leaks:
```
import tensorflow as tf
for i in range(int(1e7)):
    with tf.name_scope("first/"):
        with tf.name_scope("second"):
            pass
```
or
```
import tensorflow_probability as tfp
distr = tfp.distributions.Normal(loc=1.0, scale=2, name="foo/")
distr._name += "/"  # Force "/" because it's removed in the constructor
for i in range(int(1e7)):
    distr.sample(name="sample/")
```
It seems like ctx.name_scope is now using a C++ Safe_PyObjectPtr. I'm not sure why but maybe the reference counter is incorrect.

**Comment by Rupeshkoushik** [comment]
Hi ,can I work on this issue?

**Comment by amahendrakar** [comment]
@matthieucoquet,
> It seems using nested name_scope can lead to memory leak. For example, in my setup, the memory grows indefinitely using this code.

I did not observe any memory leaks while running this given code snippet. Although second example runs indefinitely, the memory consumption is constant. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/eb2309eed05d0cbe8cb9e873eb380bf3/45322-tf-nightly.ipynb). Thanks!

**Comment by matthieucoquet** [comment]
@amahendrakar 
I see the RAM usage going up in the colab notebook you linked (first sample, using only name_scope).

Using the tracemalloc module, I also see increasing memory usage in the colab notebook.
```
import tensorflow as tf
import tracemalloc
tracemalloc.start()

snapshot1 = tracemalloc.take_snapshot()
for i in range(int(1e7)):
    with tf.name_scope("first"):
        with tf.name_scope("second"):
            pass
snapshot2 = tracemalloc.take_snapshot()
top_stats = snapshot2.compare_to(snapshot1, 'lineno')
print("[ Top 10 differences ]")
for stat in top_stats[:10]:
    print(stat)
```

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with the latest [TF-nightly](https://colab.research.google.com/gist/amahendrakar/2e47c65e4cb30ab59a3fc83d3e288137/45322-tf-nightly.ipynb), i.e. v2.5.0-dev20201208.

However, I did not observe much increase in the memory usage with [TF v2.2](https://colab.research.google.com/gist/amahendrakar/03b1a1862050fc7013ad3e217f828d3a/45322-2-2.ipynb) and [TF v2.3](https://colab.research.google.com/gist/amahendrakar/f5b75de4bef3971db0edd28effa642a7/45322-2-3.ipynb). Please check the linked gist for reference. Thanks!

**Comment by oawxkw** [comment]
I also observed the following API aliases or similar APIs can cause the same issue in older versions of tensorflow.
Users should be cautious when using them on both CPU and GPU from tensorflow 2.4.0 (v2.4.0-rc4-71-g582c8d236cb) to 2.5.1 (v2.5.0-160-g8222c1cfc86).

- `(tf.name_scope)`, `tf.compat.v1.name_scope`

<details>
  <summary>Code to reproduce the issue in <code>tf.compat.v1.name_scope</code> in older versions</summary>

```python
import tensorflow as tf
print(tf.version.GIT_VERSION, tf.version.VERSION, flush=True)
print(tf.config.list_physical_devices(), flush=True)


import tracemalloc
tracemalloc.start()


snapshot1 = tracemalloc.take_snapshot()
for i in range(int(1e7)):
    with tf.compat.v1.name_scope("first"):
        with tf.compat.v1.name_scope("second"):
            pass
snapshot2 = tracemalloc.take_snapshot()
top_stats = snapshot2.compare_to(snapshot1, 'lineno')
print("[ Top 10 differences ]")
for stat in top_stats[:10]:
    print(stat)
```

On my GPU machine, xxx.

```text
v2.5.0-160-g8222c1cfc86 2.5.1
[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'), PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]
[ Top 10 differences ]
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py:6676: size=525 MiB (+525 MiB), count=10000000 (+10000000), average=55 B
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py:6602: size=488 B (+488 B), count=1 (+1), average=488 B
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py:6598: size=480 B (+480 B), count=1 (+1), average=480 B
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py:6605: size=472 B (+472 B), count=1 (+1), average=472 B
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py:6690: size=448 B (+448 B), count=1 (+1), average=448 B
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py:6479: size=448 B (+448 B), count=1 (+1), average=448 B
45322-5-p/tf.compat.v1.name_scope.py:37: size=440 B (+440 B), count=1 (+1), average=440 B
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py:6677: size=440 B (+440 B), count=1 (+1), average=440 B
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py:6462: size=432 B (+432 B), count=1 (+1), average=432 B
45322-5-p/tf.compat.v1.name_scope.py:35: size=424 B (+424 B), count=1 (+1), average=424 B
```

This behavior is also reproducible on my CPU machine:

```text
v2.5.0-160-g8222c1cfc86 2.5.1
[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU')]
[ Top 10 differences ]
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py:6676: size=525 MiB (+525 MiB), count=10000000 (+10000000), average=55 B
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py:6602: size=488 B (+488 B), count=1 (+1), average=488 B
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py:6598: size=480 B (+480 B), count=1 (+1), average=480 B
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py:6605: size=472 B (+472 B), count=1 (+1), average=472 B
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py:6690: size=448 B (+448 B), count=1 (+1), average=448 B
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py:6479: size=448 B (+448 B), count=1 (+1), average=448 B
45322-5-p/tf.compat.v1.name_scope.py:37: size=440 B (+440 B), count=1 (+1), average=440 B
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py:6677: size=440 B (+440 B), count=1 (+1), average=440 B
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/ops.py:6462: size=432 B (+432 B), count=1 (+1), average=432 B
45322-5-p/tf.compat.v1.name_scope.py:35: size=424 B (+424 B), count=1 (+1), average=424 B
```
</details>

It seems to be fixed in tensorflow 2.6.0 (v2.6.0-rc2-32-g919f693420e) and later versions.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
